### PR TITLE
Exposed Tangram api through event 

### LINKED ISSUE #203: Expose Tangram features
- There has been needs to expose Tangram api through mapzen.js ( #195 ) 
- Since Tangram is injected dynamically (async) , I am not sure what is the best way to expose Tangram api. 
- Currently, `setTimeout` is being used to wait for Tangram to load. This might need to be replaced with Promise to return actual Tangram layer

### PR DESCRIPTION
**Pull request by hanbyul-here**
- There has been needs for exposing Tangram feature through `mapzen.js`. Before, users loop through each layer of Leaflet and grab [a Tangram layer](https://github.com/mapzen/mapzen.js/issues/195)
- This commit exposed `tangram` wrapper object through `L.Mapzen.tangram`, specifically tangram layer through event `loaded`. You can see how to use the event at [here](https://github.com/mapzen/mapzen.js/blob/hanb/tangram-promise/src/index.html#L64). 
- Current change tries to follow Leaflet plugin's syntax. I wonder this style of syntax doesn't make it clear for users that `L.Mapzen.tangram` can't actually be used to set Tangram scene or fallback tiles. 
- @bcamper @binx I will appreciate any feedback! 
- @rmglennon letting you know about this API change.
- closes #203

### CODE REVIEW

File: src/index.html [hunk 1]
```diff
@@ -61,6 +61,11 @@
         geocoder: geocoder
       });
 
+      var tangram = L.Mapzen.tangram();
+      tangram.on('loaded', function(e) {
+        console.log(e.layer);
```
Comment (bcamper) [inline src/index.html]: Stray console log

**Comment by bcamper** [comment]
I'm a big fan of promises, but is it worth having to introduce a promise polyfill just to use them in this case? Do we have future cases in mind?

File: src/index.html (hunk 1 above)
Comment (bcamper) [inline src/index.html]: I see you have these throughout. My ingrained behavior is to not ship code with raw console logs, but perhaps this is intentional.

**Comment by hanbyul-here** [comment]
Actually, no I don't have specific plans about Promise. Probably it is not necessary for now. Thanks for pointing out @bcamper

**Comment by migurski** [comment]
Okay, @hanbyul-here and I talked about this a bit, and I think we've arrived at a slightly tighter API for this. I was confused, because I did not understand what the return value of `L.mapzen.tangram()` was doing — it’s gives you an event hook for when the scene is loaded, but it’s not immediately obvious how that works or whether events other than `'loaded'` will work.

Instead, we're thinking to take advantage of how a Leaflet map [already fires events](http://leafletjs.com/reference.html#map-events), to add a new one called `tangram-loaded`. This will make it possible write code like this:

```Javascript
L.mapzen.on('tangram-loaded', function(layer)
{
    // do stuff with layer and layer.scene!
});
```

…instead of code like this:

```Javascript
function waitForScene(map)
{
    var scene;
    
    map.eachLayer(function(layer)
    {
      // find the first layer with a Tangram scene.
      if(layer.scene)
      {
        scene = layer.scene;
        return;
      }
    });

    if(scene == undefined) {
        window.setTimeout(waitForScene, 100, map);

    } else {
        // do stuff with the scene!
    }
}

waitForScene(map);
```

**Comment by hanbyul-here** [comment]
To be clear, it will be
```
var map = L.Mapzen.map('map');
map.on('tangramloaded', function (e) {
// do something with e.tangramLayer
});
```

**Comment by bcamper** [comment]
Actually Tangram kind of has an event like this, an `init` Leaflet event which is fired after:

- The layer is added to Leaflet (e.g. `layer.addTo(map)`)
- The Tangram scene finishes loading (e.g. `layer.scene.load().then(...)`)

The Tangram scene can't be loaded until it's added to Leaflet anyway so it can get container and view information.

In hindsight, I think something more scoped to Tangram would have made sense naming-wise. 

Anyway it's possible you can use that, or put a thin wrapper around it to trigger a separate event.

**Comment by bcamper** [comment]
Sorry I should clarify that since the scene can't be loaded until the layer is added to the map, there's no scene object to manipulate before that anyway.

**Comment by migurski** [comment]
The event is called "init"?

**Comment by bcamper** [comment]
Yep

**Comment by migurski** [comment]
Huh… any way to change that in a future version of Tangram? Would it be accessible via the Leaflet map’s events?

**Comment by bcamper** [comment]
The `init` event is a Leaflet event fired on the Tangram layer object (which is a subclassed Leaflet layer so I think it's OK for it to add its own events), not on the map object itself. This works fine for people using Tangram directly, and is actually used in many (most?) of our demos:

https://github.com/tangrams/tangram/blob/master/demos/main.js#L494-L503

I think it co-exists with your new event: you can listen for the Tangram layer's event, and trigger a second map-wide event (since the point as I understand it is that mapzen.js users don't know which layer is the Tangram one).

**Comment by migurski** [comment]
So it sounds like we might listen for the `tangramloaded` event on the map to determine the active layer, and then listen for the `init` event on that layer?

**Comment by bcamper** [comment]
I think this is what you're saying (maybe @hanbyul-here can weigh in) but my understanding is that the `tangramloaded` is an event that a user of mapzen.js would listen to, to access the Tangram layer when it's available, as here: https://github.com/mapzen/mapzen.js/pull/209/files#diff-e249faefed5757034596c5096d33dab6R64

My suggestion is simply that Tangram's `init` event, which is fired on its own layer when it's done loading, is probably the best way to determine when mapzen.js should fire its own `tangramloaded` event. The Tangram `init` can replace whatever is going on [here](https://github.com/mapzen/mapzen.js/pull/209/files#diff-d5a1b3d3b699e8ed2398bfd5239262f1R20) (I'm not entirely sure of the event flow here).

So:
- Mapzen.js creates a Tangram layer and adds it to the map.
- Mapzen.js listens for the `init` event on the Tangram layer.
- When Tangram's `init` fires, mapzen.js fires `tangramloaded` on the map object. (Users wouldn't have a way to listen for Tangram's `init` event since they don't have access to the layer... the entire point of this ticket in the first place :)

Does that make sense?

**Comment by migurski** [comment]
Yes, thank you — I was imagining the opposite, this makes more sense.

**Comment by hanbyul-here** [comment]
based on what @bcamper said, current event fire can be replaced by `init` event of Tangram. (Map object is going to wrap this event, so docs don't need to be changed) I will push this version soon.

Also this whole thread makes me think it might be eventually better to give user to control `tangramLayer` rather than Map object automatically handling it.

so it can be

```
var map = L.Mapzen.map('map');
var tLayer = L.Mapzen.tangram({
scene: L.Mapzen.Refill});
tLayer.addTo(map);
```

**Comment by bcamper** [comment]
I think providing access to the Tangram layer one way or another is important, but I don't think the user should have to explicitly initialize a Tangram layer (`L.Mapzen.tangram(...)`). I think we want to make it as easy as possible to just "get a map", and ideally `L.Mapzen.map(...)` by itself should result in map without additional calls.

**Comment by migurski** [comment]
I would agree; the purpose of mapzen.js to me is to wrap up the creation of a map with all the Tangram bits pre-done. I don't think giving the user control here benefits them.

**Pull request merged**